core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java

```diff
@@ -584,6 +606,26 @@ public Builder localNodeId(String localNodeId) {
             return this;
         }
 
+        public DiscoveryNode nodeWithSameAddress(DiscoveryNode node) {
```

Comment:
I have no idea how often this is called but maybe it's worth to add the node ID as a key in the map too? ie. put multiple keys for one `DiscoveryNode` into the `nodes` map?